gcp_deepvariant_runner.py

```diff
@@ -321,6 +395,83 @@ def _get_gcs_relative_path(gcs_path):
 
 
+def _get_bam_category(pipeline_args):
+  """Returns the category that input BAM files belongs to."""
+  bam_size = _get_gcs_object_size(pipeline_args.bam)
+  if bam_size == 0:
+    logging.warning('Size of input bam file is 0.')
+
+  is_wes = pipeline_args.model.find(_WES_STANDARD) != -1
+  is_wgs = pipeline_args.model.find(_WGS_STANDARD) != -1
+
+  if is_wes:
+    if bam_size < _WES_LARGE_THR:
+      return BamCategories.WES_SMALL
+    else:
+      return BamCategories.WES_LARGE
+
+  if is_wgs:
+    if bam_size < _WGS_SMALL_THR:
+      return BamCategories.WGS_SMALL
+    elif bam_size > _WGS_LARGE_THR:
+      return BamCategories.WGS_LARGE
+    else:
+      return BamCategories.WGS_MEDIUM
+
+
+def _set_args_based_on_bam_category(bam_category, pipeline_args):
+  """Sets pipeline args using the given bam_category and default flag values."""
+  default_flags = _DEFAULT_FLAGS[bam_category]
+  pipeline_args.shards = (
+      default_flags['make_examples_workers'] *
+      default_flags['make_examples_cores_per_worker'])
+  pipeline_args.make_examples_workers = (default_flags['make_examples_workers'])
+  pipeline_args.make_examples_cores_per_worker = (
+      default_flags['make_examples_cores_per_worker'])
+  pipeline_args.make_examples_ram_per_worker_gb = (
+      default_flags['make_examples_cores_per_worker'] * _RAM_PER_CORE)
+  pipeline_args.make_examples_disk_per_worker_gb = (
+      _MAKE_EXAMPLES_DISK_PER_WORKER)
+  if 'gpu' in default_flags:
+    pipeline_args.gpu = default_flags['gpu']
+    pipeline_args.call_variants_workers = (
+        default_flags['call_variants_workers'])
+    pipeline_args.call_variants_cores_per_worker = (
+        default_flags['call_variants_cores_per_worker'])
+    pipeline_args.call_variants_ram_per_worker_gb = (
+        default_flags['call_variants_cores_per_worker'] * _RAM_PER_CORE)
+    pipeline_args.call_variants_disk_per_worker_gb = (
+        _CALL_VARIANTS_DISK_PER_WORKER)
+  elif 'tpu' in default_flags:
+    pipeline_args.tpu = default_flags['tpu']
+    pipeline_args.gke_cluster_zone = pipeline_args.zones[0]
+  else:
+    raise ValueError('Either gpu or tpu is needed for default flag settings.')
+  # Following flags are independent of BAM file category.
+  pipeline_args.preemptible = True
+  if pipeline_args.gvcf_outfile:
+    pipeline_args.postprocess_variants_disk_gb = _POSTPROCESS_VARIANTS_DISK_GVCF
+
+
+def _set_computational_flags_based_on_bam_size(pipeline_args):
+  """Automatically sets computational flags based on size of input BAM file."""
+  # First validating all necessary flags are present.
+  if not (pipeline_args.docker_image and pipeline_args.docker_image_gpu):
+    raise ValueError('both --docker_image and --docker_image_gpu must be '
+                     'provided with --set_optimized_flags_based_on_bam_size')
+  is_wes = pipeline_args.model.find(_WES_STANDARD) != -1
+  is_wgs = pipeline_args.model.find(_WGS_STANDARD) != -1
+  if is_wes == is_wgs:
```

Comment:
Original comments

"
nmousavi@
I don't think this check belongs to here. Maybe move it to where we validate flags.

Anyway, I think it should be as below

if not is_wes and not is_wgs:
....

It's possible to have both is_wes and is_wgs true, thought would be a bug.

  @samanvp
Here we need is_wes XOR is_wgs meaning that exactly one of those two must be True.

And for placement, note that _set_computational_flags_based_on_bam_size is run before validating flags. This is a conscious decision since we should first run this piece of code to set all the flags automatically and then we validate flags; regardless if they were set manually by user or automatically.

Please let me know if you think this needs to change.
"

Reply:
Sorry I was not clear on the first point. What I meant is that the two cases (not is_wes and not is_wgs) and (is_wes and is_wgs) are different, and you may want to show different error.

re the second point:
I am not sure why flag validation has to be done after setting flags set_computational_flags...()? Some flags have to be validated before setting computational flags, and that's why this two concerns are mixed here. Can we have all flags validated first, and then set computational flags? Or you think once our code set computational flags, we need to validate it? why? In that I am proposing, these two concerns (flag validation and setting computational flags) are not mixed.